### PR TITLE
Dolci/final setup

### PR DESCRIPTION
**Pull request by Ig-dolci**
The main modifications are in the **StorageType**. Previously, the **StorageType.WORK** was the type of storage used either for the forward restarting or in the adjoint computation. In this branch, I added **StorageType.FWD_RESTART** and **StorageType.ADJ_DEPS**. Hence, the action  Copy(n, StorageType.RAM, StorageType.FWD_RESTART) means the copy of the checkpoint data from RAM to a "variable" that holds the initial condition for the forward recomputation. When we have the action Forward(n0, n1, True, False, StorageType.FWD_RESTART), it means to store the forward data in the "variable" that holds the data used to initialise the forward solver. Whereas Forward(n0, n1, False, True, StorageType.ADJ_DEPS) means storing the forward data in a "variable" that holds the forward data required for the adjoint computation.

### CODE REVIEW

File: checkpoint_schedules/none_revolver.py [hunk 1]
```diff
@@ -1,27 +1,49 @@
...
+    are stored in memory.
+
+    Parameters
+    ----------
+    write_ics : bool
+        Indicate whether to store the forward restart data for all steps.
+    storage_ics: enum
+        Indicate the storage type of the forward restart data. 
+        This atributte is checked only if the user desires to save the forward 
+        restart.
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:33]: Unused

File: checkpoint_schedules/none_revolver.py [hunk 2]
```diff
@@ -1,27 +1,49 @@
...
 
+    Notes
+    -----
     Online, unlimited adjoint calculations permitted.
+
+    `write_ics` is always `False` for this schedule by considering that storing
+    the forward restart data is unnecessary by this schedule, as there is no
+    need to recompute the forward solver while time advancing the adjoint
+    solver.
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:42]: Not needed

File: checkpoint_schedules/none_revolver.py [hunk 3]
```diff
@@ -58,13 +79,12 @@ def is_exhausted(self):
         return False
 
     def uses_storage_type(self, storage_type):
-        """Check if a given storage type is used in this schedule.
+        """Check the storage type.
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:82]: Should be more descriptive

File: checkpoint_schedules/none_revolver.py [hunk 4]
```diff
@@ -73,11 +93,103 @@ def uses_storage_type(self, storage_type):
         """
 
         return storage_type == self._storage
+    
+
+class SingleDiskStorageSchedule(CheckpointSchedule):
+    """A checkpointing schedule where all adjoint dependencies
+    are stored in disk.
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:100]: on disk

File: checkpoint_schedules/none_revolver.py [hunk 5]
```diff
@@ -73,11 +93,103 @@ def uses_storage_type(self, storage_type):
...
+
+    Notes
+    -----
+    Online, unlimited adjoint calculations permitted.
+
+    Parameters
+    ----------
+    storage : enum
+        Indicate that the execution should stores in memore all foward restart 
+        data and non-linear dependency data.
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:110]: Unused

File: checkpoint_schedules/none_revolver.py [hunk 6]
```diff
@@ -73,11 +93,103 @@ def uses_storage_type(self, storage_type):
...
+
+    Notes
+    -----
+    Online, unlimited adjoint calculations permitted.
+
+    `write_ics` is always `False` for this schedule by considering that storing
+    the forward restart data is unnecessary by this schedule, as there is no
+    need to recompute the forward solver while time advancing the adjoint
+    solver.
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:119]: Not needed

File: checkpoint_schedules/none_revolver.py [hunk 7]
```diff
@@ -73,11 +93,103 @@ def uses_storage_type(self, storage_type):
...
+    """
+
+    def __init__(self, move_data=False):
+        self._move_data = move_data
+        self._storage = StorageType.ADJ_DEPS
+        super().__init__()
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:125]: `__init__` should be called before setting attributes

File: checkpoint_schedules/none_revolver.py [hunk 8]
```diff
@@ -73,11 +93,103 @@ def uses_storage_type(self, storage_type):
...
+    -----
+    Online, unlimited adjoint calculations permitted.
+
+    `write_ics` is always `False` for this schedule by considering that storing
+    the forward restart data is unnecessary by this schedule, as there is no
+    need to recompute the forward solver while time advancing the adjoint
+    solver.
+    """
+
+    def __init__(self, move_data=False):
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:122]: `move_data` undocumented

File: checkpoint_schedules/none_revolver.py [hunk 9]
```diff
@@ -73,11 +93,103 @@ def uses_storage_type(self, storage_type):
...
+            n1 = n0 + sys.maxsize
+            self._n = n1
+            yield Forward(n0, n1, False, True, StorageType.DISK)
+
+        yield EndForward()
+
+        for i in range(self._max_n, 0, -1):
+            if self._r < self._max_n:
+                # Reverse
+                if self._move_data is True:
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:146]: `if self._move_data`. It's unsafe to test using `is True` or `is False`, e.g. `np.int64(1) == 1 is True` is `False`.

File: checkpoint_schedules/none_revolver.py [hunk 10]
```diff
@@ -73,11 +93,103 @@ def uses_storage_type(self, storage_type):
...
+
+                self._r = 0
+                yield EndReverse(False)
+            else:
+                raise RuntimeError("Invalid checkpointing state")
+            self._r += 1
+
+    @property
+    def is_exhausted(self):
+        return False
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:163]: Schedule should exhaust if `move_data` is `True`.

File: checkpoint_schedules/none_revolver.py [hunk 11]
```diff
@@ -73,11 +93,103 @@ def uses_storage_type(self, storage_type):
...
+
+    def uses_storage_type(self, storage_type):
+        """Check if the storage is in memory.
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:166]: Should be more descriptive

File: checkpoint_schedules/none_revolver.py [hunk 12]
```diff
@@ -73,11 +93,103 @@ def uses_storage_type(self, storage_type):
...
+
 
 class NoneCheckpointSchedule(CheckpointSchedule):
     """A checkpointing schedule for the case where no adjoint calculation is
     performed.
 
+    Parameters
+    ----------
+    _exhausted : bool
+        Indicate that the execution is exhausted.
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:189]: Unused / not a parameter

File: checkpoint_schedules/none_revolver.py [hunk 13]
```diff
@@ -103,21 +215,22 @@ def _iterator(self):
 
     @property
     def is_exhausted(self):
+        """Check if the execution is exhausted.
+        """
         return self._exhausted
 
     def uses_storage_type(self, storage_type):
         """Check the storage type.
```
Comment (jrmaddison) [inline checkpoint_schedules/none_revolver.py:223]: Should be more descriptive

File: checkpoint_schedules/schedule.py [hunk 14]
```diff
@@ -1,5 +1,7 @@
 #!/usr/bin/env python3
 # -*- coding: utf-8 -*-
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:2]: Not needed in module files

File: checkpoint_schedules/schedule.py [hunk 15]
```diff
@@ -19,48 +21,71 @@
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:21]: Extra new line

File: checkpoint_schedules/schedule.py [hunk 16]
```diff
@@ -19,48 +21,71 @@
 
 
 class StorageType(Enum):
-    """
-    RAM : It is the first type of checkpoint storage.
+    """ Enums to indicate the storage type of the checkpoint data.
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:24]: `StorageType` is the `Enum`, `RAM`, `DISK` etc are members

File: checkpoint_schedules/schedule.py [hunk 17]
```diff
@@ -19,48 +21,71 @@
...
+
+    RAM : A enum to indicate the storage of the forward checkpoint data in memory.
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:26]: Should just identify the location, e.g. might not store forward restart data

File: checkpoint_schedules/schedule.py [hunk 18]
```diff
@@ -19,48 +21,71 @@
...
-    DISK : It is the second type of checkpoint storage.
+    DISK : A enum to indicate the storage of the forward checkpoint data in RAM.
     
-    WORK : ...
+    FWD_RESTART : A enum to indicate the storage of the forward checkpoint data
+    in a local used to initilise the forward solver.
 
-    NONE : Indicate that there is no specific storage location defined 
+    ADJ_DEPS : A enum to indicate the storage of the forward checkpoint data
+    in a local used for the adjoint computation.
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:34]: Forward restart and adjoint dependency data are not in general distinct, and so these should not refer to different storage locations

File: checkpoint_schedules/schedule.py [hunk 19]
```diff
@@ -19,48 +21,71 @@
...
     """
     RAM = 0
     DISK = 1
-    WORK = -1
+    FWD_RESTART = 2
+    ADJ_DEPS = 3
     NONE = None
 
 
 class StepType(IntEnum):
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:46]: If this is used for tabulation then I think this should appear in utils.py

File: checkpoint_schedules/schedule.py [hunk 20]
```diff
@@ -19,48 +21,71 @@
...
+    """A checkpoint action base class.
+
     Attributes
     ----------
     *args : Any
-        The *args correspond to the arguments of the checkpoint scheduled actions: 
-        `Forward`, `Reverse`, `Copy`, `EndForward`, `EndReverse`.
-    
+        The *args correspond to the arguments of the checkpoint scheduled 
+        actions: `Forward`, `Reverse`, `Copy`, `EndForward`, `EndReverse`.
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:82]: Missing `Move`. Maybe do not list subclasses, as it's difficult to maintain a list like this.

File: checkpoint_schedules/schedule.py [hunk 21]
```diff
@@ -19,48 +21,71 @@
...
+
     See Also
     --------
     :class:`Forward`, :class:`Reverse`, :class:`Copy`, :class:`EndForward`,
     :class:`EndReverse`.
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:87]: Similarly I'd remove this, as it's difficult to maintain

File: checkpoint_schedules/schedule.py [hunk 22]
```diff
@@ -73,38 +98,36 @@ def __eq__(self, other):
 
 
 class Forward(CheckpointAction):
-    """This action indicates the advancement of the forward solver 
-    and configures the intermediate storage.
+    """This action indicates the forward advancement.
+    This action also configures the intermediate storage.
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:102]: 'intermediate storage' is a tlm_adjoint concept

File: checkpoint_schedules/schedule.py [hunk 23]
```diff
@@ -534,7 +480,7 @@ def is_running(self):
         -------
         bool
             Indicate whether the :class:`CheckpointSchedule` schedule has an 
-            iterator object. Do not finalise the checkpoint schedule if ``True``.
+            iterator object. Do not finalise the checkpoint schedule if `'True'`.
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:483]: Running schedules can be finalized. `is_running` returns whether at least one action has been yielded.

File: checkpoint_schedules/schedule.py [hunk 24]
```diff
@@ -387,15 +319,17 @@ class EndReverse(CheckpointAction):
     """A checkpointing action which indicates the end of an adjoint
     calculation.
     """
-    
+
```
Comment (jrmaddison) [inline checkpoint_schedules/schedule.py:322]: Should add a `__init__` so that supplying arguments raises an exception

**Comment by jrmaddison** [comment]
I've looked at and commented on none_revolver.py and schedule.py for now and suggested changes that might apply elsewhere. The code should also be linted with flake8.

File: checkpoint_schedules/none_revolver.py [hunk 25]
```diff
@@ -1,27 +1,49 @@
 #!/usr/bin/env python3
 # -*- coding: utf-8 -*-
-"""..."""
+
+"""This module contains the checkpointing schedules for the cases where no 
+revolver algorithm are used.
```
[the remark this answers is not in the document]
Comment (Ig-dolci) [reply checkpoint_schedules/none_revolver.py:5]: Maybe `basic_schedules.py` or `trivial_schedules.py`?

File: checkpoint_schedules/none_revolver.py (hunk 9 above)
Comment (Ig-dolci) [reply checkpoint_schedules/none_revolver.py:146]: Yes. 🤦🏻‍♀️

Comment (Ig-dolci) [reply checkpoint_schedules/none_revolver.py:146]: Yes. 🤦🏻‍♀️

File: checkpoint_schedules/none_revolver.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (jrmaddison) [reply checkpoint_schedules/none_revolver.py:5]: Sounds good, I prefer basic_schedules.py

**Pull request merged**